azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java

```diff
@@ -902,37 +913,36 @@ private void prepareJobProperties(final ExecutableNode node) throws IOException
       props = jobSource;
     }
 
-    if (this.azkabanProps.getBoolean(
-        ConfigurationKeys.EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED, false)) {
-      // Flow override props are configured to also override existing job props
-      // =>
-      // 5. If there are any runtime flow overrides, we apply them now.
+    // 5. If configured, runtime properties override also existing .job props & output props
```

Comment:
I think we still don't quite understand each other. Let me try to show you what I mean with some code.
Here is how I would implement the overwrites:
-remove lines 894 to 897 
```
      if (!isOverrideExistingEnabled()) {
         // if there are more specific runtime props, those override the flow level props
         flowProps = applyRuntimeProperties(node, runtimeProperties, flowProps);
       }
``` 
-remove lines 916 to 926
```
   // 5. If configured, runtime properties override also existing .job props & output props
   if (isOverrideExistingEnabled()) {
       // 5.1. apply flow level runtime props
       final Map<String, String> flowParam =
           this.flow.getExecutionOptions().getFlowParameters();
       if (flowParam != null && !flowParam.isEmpty()) {
         props = new Props(props, flowParam);
       }
       // 5.2. apply node-specific runtime props
       props = applyRuntimeProperties(node, runtimeProperties, props);
    }
```    
-remove methods `applyRuntimeProperties` and `isOverrideExistingEnabled`

-add below line 914
```
    // 4. The job source.
    final Props jobSource = loadJobProps(node);
    if (jobSource != null) {
       jobSource.setParent(props);
       props = jobSource;
    }

//new code
     if (this.azkabanProps.getBoolean(
         ConfigurationKeys.EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED, false)) {
     	Node currentNode = node;
     	Props overrideProps = null;
     	do {
     		if (runtimeProperties.containsKey(currrentNode.getNestedId())) {
     			Props currentOverrides = new Props(null, runtimeProperties.get(currentNode.getNestedId()));
     			if (overrideProps == null) {
     				overrideProps = currentOverrides;
     			} else {
     				overrideProps.setEarliestAncestor(currentOverrides);
     			}
    	 	}
    	 	currentNode = currentNode.getParentFlow();
     	} while (currentNode != null);
     	overrideProps.setEarliestAncestor(props);
     	props = overrideProps;
     } else {
     	if (runtimeProperties.containsKey(node.getNestedId())) {
     		props = new Props(props, runtimeProperties.get(node.getNestedId()));
     	}
     }
  // end of new code
``` 
This will ensure: 
 1-overwrites specific to this node have the highest precedence 
 2-overwrites on all parent flows including the root flow will follow in precedence if EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED is true, otherwise they will have the default precedence as they are available in the parent's input properties.

Reply:
That code snippet didn't compile as it is. To make it compile I made these changes:

```
-      	Node currentNode = node;
+      	ExecutableNode currentNode = node;
```

```
-      		if (runtimeProperties.containsKey(currrentNode.getNestedId())) {
+      		if (runtimeProperties.containsKey(currentNode.getNestedId())) {
```

However, the result of FlowRunnerPropertyResolutionTest after your suggested changes is surprising:
- tests with EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=false are still passing
- tests with EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=true are failing

I thought you would be mostly concerned of the default behavior, ie. EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=false.

Also:

> 1-overwrites specific to this node have the highest precedence

I have implemented it to be like that only if the conf flag is set to also override existing. I understood that otherwise runtime props are not supposed to have _highest_ precedence; if the job itself defines the same property, then the runtime property doesn't matter. We discussed this quite extensively, right?

> 2-overwrites on all parent flows including the root flow will follow in precedence if EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED is true, otherwise they will have the default precedence as they are available in the parent's input properties.

I think I have implemented it like that, but I could be missing something of course.

I would kindly ask if you could review once more the logic that I have implemented currently. Is there an example of some case in practice where the result is wrong? To avoid possibility for confusion entirely, could you modify the unit test locally & share it as an actual commit via git?

Reply:
@ypadron-in, would you have time to check this? ^

Reply:
Apologies for the delays!
Here is the commit you requested: https://github.com/ypadron-in/azkaban/commit/6779d65e35193925ee1a0d18a3e434753bd8c4be. It includes props resolution logic and tests.
You were right, the code snippet I shared above was not analyzing possible overrides to the root flow node.

Reply:
Please, check my comment on https://github.com/ypadron-in/azkaban/commit/6779d65e35193925ee1a0d18a3e434753bd8c4be#r51035789 too.

Reply:
Thanks a lot for that!

Your test revealed this actual bug in my code:
Node-specific override didn't win over `.properties` that are on the same level in folder hierarchy as the job itself. That was the case because `loadJobProps(node)` also loads the `.properties` and I had not realized that (I somehow assumed that it would only load the `.job` itself). In the unit test my bug was revealed by a property that's set in `test/execution-test-data/execpropstest/subdir/shared.properties`.

Otherwise I had a misunderstanding: I thought that props that are defined by the job itself are always highest priority. I also thought the same about output props (coming from the previous node), but that must be my misunderstanding. According to the asserts in your version of the unit test, an actual node-specific override simply wins _always_.

With those rules, it was easy to fix my code to satisfy all tests – see my latest commit in this PR, which is basically a one-liner.